// @sudoku/stores/hints.js
import { settings } from './settings';
import { writable } from 'svelte/store';

export const usedHints = writable(0);

/**
 * 提示模式下的候选数展示缓存
 * 键为 "x,y" 字符串，值为候选数字数组（如 [2,5,7]）
 * 写入后触发 Board 组件在对应格子渲染 3x3 迷你候选数网格
 */
export const hintCandidates = writable({});

/** 当前棋盘是否存在唯一推定值的格子，用于控制"下一步"按钮 disabled 状态 */
export const hasNextHint = writable(false);

function createHints() {
	let defaultHints = Infinity;

	const hints = writable(Infinity);

	settings.subscribe(($settings) => {
		if ($settings.hintsLimited) {
			defaultHints = $settings.hints;
			hints.update($hints => {
				if ($hints > $settings.hints) return $settings.hints;

				return $hints;
			})
		} else {
			defaultHints = Infinity;
			hints.set(Infinity);
		}
	});

	return {
		subscribe: hints.subscribe,

		useHint() {
			hints.update($hints => {
				if ($hints > 0) {
					usedHints.update($usedHints => $usedHints + 1);
					return $hints - 1;
				}

				return 0;
			});
		},

		reset() {
			hints.set(defaultHints);
			usedHints.set(0);
			hintCandidates.set({});
			hasNextHint.set(false);
		}
	};
}

export const hints = createHints();
